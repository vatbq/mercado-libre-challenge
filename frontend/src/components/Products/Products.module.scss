@import "styles/_global";
@import "styles/_colors";

.root {
  display: flex;
  flex-direction: column;
  padding-left: $layout-padding-horizontal;
  padding-right: $layout-padding-horizontal;
}

.productsContainer {
  display: grid;
  grid-template-columns: 1fr;
  background-color: $white;
  border-radius: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  margin-top: 1rem;
}
