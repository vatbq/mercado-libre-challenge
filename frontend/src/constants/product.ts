export const ProductCondition = {
  New: "new",
  Used: "used",
} as const
