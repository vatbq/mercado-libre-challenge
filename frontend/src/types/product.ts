export interface Product {
  id: string;
  title: string;
  price: {
    currency: string;
    amount: string;
    decimals: string;
  };
  picture: string;
  condition: string;
  free_shipping: boolean;
  sold_quantity: number;
  description: string;
  categories: string[];
}
