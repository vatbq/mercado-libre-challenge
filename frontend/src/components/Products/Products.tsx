import { Product } from "types/product";
import ProductCard from "components/ProductCard";
import Breadcrumb from "components/Breadcrumb";

import classes from "./Products.module.scss";

interface ProductsProps {
  products: Product[];
  categories: string[];
}

const Products = ({ products, categories }: ProductsProps) => {
  const breadcrumbItems = categories.map((category) => ({ item: category }));

  return (
    <div className={classes.root}>
      {breadcrumbItems.length > 0 && <Breadcrumb items={breadcrumbItems} />}
      <div className={classes.productsContainer}>
        {products.map((product) => (
          <ProductCard key={product.id} product={product} />
        ))}
      </div>
    </div>
  );
};

export default Products;
