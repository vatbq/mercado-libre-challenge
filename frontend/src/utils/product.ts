import { Product } from "types/product";

const numberFormatter = new Intl.NumberFormat("en-US", {
  style: "decimal",
  minimumFractionDigits: 0,
  maximumFractionDigits: 2,
});

export const getProductPrice = ({ price }: Product): string => {
  const decimals = price.decimals ? parseFloat(price.decimals) : 0;
  return numberFormatter.format(parseFloat(`${price.amount}.${decimals}`));
};
