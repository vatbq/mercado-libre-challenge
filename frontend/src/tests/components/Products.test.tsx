import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Products from "components/Products";
import { productsMock } from "tests/factories/product";

jest.mock("components/ProductCard", () => () => <div>ProductCard</div>);
jest.mock("components/Breadcrumb", () => () => <div>Breadcrumb</div>);

const categoriesMock = ["Electronics", "Gadgets"];

describe("<Products />", () => {
  it("renders ProgramCard unchanged", () => {
    const { container } = render(
      <Products products={productsMock} categories={categoriesMock} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders Breadcrumb component when categories are provided", () => {
    render(<Products products={[]} categories={categoriesMock} />);

    expect(screen.getByText("Breadcrumb")).toBeInTheDocument();
  });

  it("does not render Breadcrumb component when no categories are provided", () => {
    render(<Products products={[]} categories={[]} />);

    expect(screen.queryByText("Breadcrumb")).not.toBeInTheDocument();
  });

  it("renders the correct number of ProductCard components", () => {
    render(<Products products={productsMock} categories={categoriesMock} />);

    expect(screen.getAllByText("ProductCard")).toHaveLength(
      productsMock.length,
    );
  });
});
