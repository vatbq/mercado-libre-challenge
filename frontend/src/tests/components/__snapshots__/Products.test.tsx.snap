// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> renders ProgramCard unchanged 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      Breadcrumb
    </div>
    <div
      class="productsContainer"
    >
      <div>
        ProductCard
      </div>
      <div>
        ProductCard
      </div>
    </div>
  </div>
</div>
`;
